### DISCUSSION TITLE
Pair of Hetero Graphs

### DISCUSSION THREAD

**Discussion by tehranixyz** [pyg-team/pytorch_geometric#4768] [Q&A] [answered]
Hi,
I'm having difficulties creating a HeteroData object that stores two heterogeneous graphs.
 
The [batching page](https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html#pairs-of-graphs) provides nice details on how to create a new class based on the [Data class](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data) to store 2 graphs and also overwrite the [__inc__](https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.Data.__inc__) function to ensure the batching is working properly.
However, in my case, I have pairs of heterogeneous graphs that I want to predict their similarity.
I have around 200K pairs of graphs for each pair I have label 1 or 0 showing that the two graphs are similar or not.
The node types and edge types are the same across the graphs. So all graphs have the same node types and edge types.
Since my heterogeneous graphs are not very large, I intend to use [DataLoader](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.DataLoader) for batching.
However, I don't know how to override  __cat_dim__ and __inc__ for my HeteroData-based class.
Let's say for the first graph we have:

```
data['nodes'].x_a = ...
data['node', 'control', 'node'].edge_index_a = ...
data['node', 'data', 'node'].edge_index_a = ...
data['node', 'call', 'node'].edge_index_a = ...
data['node', 'control', 'node'].edge_attr_a = ...
data['node', 'data', 'node'].edge_attr_a = ...
data['node', 'call', 'node'].edge_attr_a = ...
```

For the second graph, we have

```
data['nodes'].x_b = ...
data['node', 'control', 'node'].edge_index_b = ...
data['node', 'data', 'node'].edge_index_b = ...
data['node', 'call', 'node'].edge_index_b = ...
data['node', 'control', 'node'].edge_attr_b = ...
data['node', 'data', 'node'].edge_attr_b = ...
data['node', 'call', 'node'].edge_attr_b = ...
```

and the label is:
`data.y = torch.tensor([1])`

How can I create a class based on HeteroData and store this information?
Thanks

**Comment by rusty1s** [comment, marked as answer]
I think you should avoid integrating `PairData` with `HeteroData` objects, this is likely to become a nightmare to maintain. Instead, you should have a `PairDataset` that returns a tuple of `HeteroData` objects:
```python
class MyDataset(...):
    ...

    def __getitem__(self, idx):
        return self.dataset1[idx], self.dataset2[idx]
```
You can then use this dataset as part of `torch_geometric.loader.DataLoader`:
```
loader = DataLoader(my_dataset, ...)

for batch1, batch2 in loader:
    pass
```
WDYT?

**Reply by tehranixyz** [reply]
That worked flawlessly. Thanks for the suggestion!

**Comment by sssleder** [comment]
I'm glad this got answered, but I'm stubborn and want to do it the hard way. Mostly so that I can understand the collating that is going on with the `HeteroData()` class but also because I spent so much time working on this functionality before I found this discussion.

I'm stuck at the same issues with modifying `HeteroData()`'s `__inc__()` function for batching to properly collate with the following error:
```
File ~/.conda/envs/from_pygcu/lib/python3.11/site-packages/torch_geometric/data/collate.py:322, in <listcomp>(.0)
    319 def get_incs(key, values: List[Any], data_list: List[BaseData],
    320              stores: List[BaseStorage]) -> Tensor:
    321     repeats = [
--> 322         data.__inc__(key, value, store)
    323         for value, data, store in zip(values, data_list, stores)
    324     ]
    325     if isinstance(repeats[0], Tensor):
    326         repeats = torch.stack(repeats, dim=0)

File ~/.conda/envs/from_pygcu/lib/python3.11/site-packages/torch_geometric/data/hetero_data.py:349, in HeteroData.__inc__(self, key, value, store, *args, **kwargs)
    347     return int(value.max()) + 1
    348 elif isinstance(store, EdgeStorage) and 'index' in key:
--> 349     return torch.tensor(store.size()).view(2, 1)
    350 else:
    351     return 0

RuntimeError: Could not infer dtype of NoneType
```
For some reason `size()` of `EdgeStorage` gives me `[None, None]`, I figure it is a side affect `num_nodes` not being properly set, but a nudge in the right direction would be helpful. I figure there is some way to associate an `EdgeStorage` instance with a `NodeStorage` instance...





As a random aside: Ali if you see this it's been a very long time and I'm glad you figured out how to parse ASTs for hetero GNNs. Also please give Dr. Jannesari my apologies for disappearing after the Euromicro paper.
